apps/libsignpost/port_signpost_tock.c

```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+// Should be a linked list of callbacks and pins, but for now just keep track
+// of a single interrupt callback
+port_signpost_callback gpio_interrupt_cb;
+static void port_signpost_gpio_interrupt_callback(
+        __attribute__ ((unused)) int pin,
+        __attribute__ ((unused)) int state,
+        __attribute__ ((unused)) int unused,
+        __attribute__ ((unused)) void* callback_args) {
+    gpio_interrupt_cb(0);
+}
+
+static uint8_t master_write_buf[I2C_MAX_LEN];
+static uint8_t slave_read_buf[I2C_MAX_LEN];
+
+//This function is called upon signpost initialization
+//You should use it to set up the i2c interface
+int port_signpost_init(uint8_t i2c_address) {
+    int rc = 0;
+    rc = i2c_master_slave_set_master_write_buffer(master_write_buf, I2C_MAX_LEN);
+    if (rc < 0) return rc;
+    rc = i2c_master_slave_set_slave_read_buffer(slave_read_buf, I2C_MAX_LEN);
+    if (rc < 0) return rc;
+    rc = i2c_master_slave_set_slave_address(i2c_address);
+    if (rc < 0) return rc;
+    return i2c_master_slave_set_callback(i2c_master_slave_callback, NULL);
+}
+
+//This function is a blocking i2c send call
+//it should return the length of the message successfully sent on the bus
+//If the bus returns an error, use the appropriate error code
+//defined in this file
+int port_signpost_i2c_master_write(uint8_t dest, uint8_t* buf, size_t len) {
+    int rc = 0;
+    memcpy(master_write_buf, buf, len);
+    master_write_yield_flag = false;
+    rc = i2c_master_slave_write(dest, len);
+    if (rc < 0) return rc;
+
+    yield_for(&master_write_yield_flag);
+    return master_write_len_or_rc;
+}
+
+//This function sets up the asynchronous i2c receive interface
+//When this function is called start listening on the i2c bus for
+//The address specified in init
+//Place data in the buffer no longer than the max len
+int port_signpost_i2c_slave_listen(port_signpost_callback cb, uint8_t* buf, size_t max_len) {
+    int rc = 0;
+    rc = i2c_master_slave_set_slave_write_buffer(buf, max_len);
+    if (rc < 0) return rc;
+    slave_write_cb = cb;
+    return i2c_master_slave_listen();
+}
+
+int port_signpost_i2c_slave_read_setup(uint8_t *buf, size_t len) {
+    memcpy(slave_read_buf, buf, len);
+    return i2c_master_slave_enable_slave_read(len);
+}
+
+//These functions are used to control gpio outputs
+int port_signpost_gpio_enable_output(int pin) {
+    return gpio_enable_output(pin);
+}
+
+int port_signpost_gpio_set(int pin) {
+    return gpio_set(pin);
+}
+
+int port_signpost_gpio_clear(int pin) {
+     return gpio_clear(pin);
+}
+
+int port_signpost_gpio_read(int pin) {
+     return gpio_read(pin);
+}
+
+//This function is used to get the input interrupt for the falling edge of
+//mod-in
+int port_signpost_gpio_enable_interrupt(int pin, InputMode input_mode, InterruptMode interrupt_mode, port_signpost_callback cb) {
+    int rc = 0;
+    gpio_interrupt_cb = cb;
+    rc = gpio_interrupt_callback(port_signpost_gpio_interrupt_callback, NULL);
```

Comment:
I had to read this several times before I realized that you weren't setting and immediately calling the gpio interrupt callback (I kept reading `cb` and `callback` as the same I guess). I think it would help readability to rename the local global state something meaningful/clarifying